solr/core/src/test/org/apache/solr/handler/component/CustomSearchHandlerTest.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.handler.component;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.solr.client.solrj.SolrQuery;
+import org.apache.solr.client.solrj.request.CollectionAdminRequest;
+import org.apache.solr.client.solrj.request.QueryRequest;
+import org.apache.solr.client.solrj.request.UpdateRequest;
+import org.apache.solr.client.solrj.response.QueryResponse;
+import org.apache.solr.cloud.AbstractDistribZkTestBase;
+import org.apache.solr.cloud.ConfigRequest;
+import org.apache.solr.cloud.SolrCloudTestCase;
+import org.apache.solr.common.util.NamedList;
+import org.apache.solr.request.SolrQueryRequest;
+import org.apache.solr.response.SolrQueryResponse;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class CustomSearchHandlerTest extends SolrCloudTestCase {
+
+  /** A custom search handler that uses a custom response builder. */
+  public static class CustomSearchHandler extends SearchHandler {
+    @Override
+    protected ResponseBuilder newResponseBuilder(
+        SolrQueryRequest req, SolrQueryResponse rsp, List<SearchComponent> components) {
+      return new CustomResponseBuilder(req, rsp, components);
+    }
+  }
+
+  /** A custom response builder that uses a custom shards info container. */
+  private static class CustomResponseBuilder extends ResponseBuilder {
+
+    public CustomResponseBuilder(
+        SolrQueryRequest req, SolrQueryResponse rsp, List<SearchComponent> components) {
+      super(req, rsp, components);
+    }
+
+    @Override
+    protected ShardsInfoContainer newShardsInfoContainer() {
+      return new CustomShardsInfoContainer();
+    }
+  }
+
+  /**
+   * A container for recording shard to shard info mappings, retaining only the "shardAddress" part
+   * of the shard info.
+   *
+   * @see org.apache.solr.common.params.ShardParams#SHARDS_INFO
+   */
+  private static class CustomShardsInfoContainer extends ShardsInfoContainer {
```

Comment:
1/3 The `CustomResponseBuilder` and `CustomShardsInfoContainer` could in principle be done as inner classes of the `CustomSearchHandler` but like this seems clearer.